File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/Mask.java
package com.vicmikhailau.maskededittext;

import java.util.ArrayList;
import java.util.List;


public class Mask {
    private String mRawMaskString;
    private List<MaskCharacter> mMask;
    private MaskCharacterFabric mFabric;
    private List<MaskCharacter> mPrepopulateCharacter;

    public Mask() {
        mFabric = new MaskCharacterFabric();
    }

    public Mask(String fmtString) {
        this();
        mRawMaskString = fmtString;
        mMask = buildMask(mRawMaskString);
    }

    public String getFormatString() {
        return mRawMaskString;
    }

    public int size() {
        return mMask.size();
    }

    public MaskCharacter get(int index) {
        return mMask.get(index);
    }

    public boolean isValidPrepopulateCharacter(char ch){
        for (MaskCharacter maskCharacter : mPrepopulateCharacter){
            if (maskCharacter.isValidCharacter(ch)) {
                return true;
            }
        }
        return false;
    }


    private List<MaskCharacter> buildMask(String fmtString) {
        List<MaskCharacter> result = new ArrayList<>();
        mPrepopulateCharacter = new ArrayList<>();
        for (char ch : fmtString.toCharArray()) {
            MaskCharacter maskCharacter =  mFabric.buildCharacter(ch);
            if (maskCharacter.isPrepopulate()) {
                mPrepopulateCharacter.add(maskCharacter);
            }
            result.add(maskCharacter);
        }
        return result;
    }


    public IFormattedString getFormattedString(String value) {
        return new FormattedString(this, value);
    }



}

File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/Mask.kt
package com.vicmikhailau.maskededittext

import java.util.*


class Mask() {
    lateinit var formatString: String
    private lateinit var mMask: List<MaskCharacter>
    private val mFabric: MaskCharacterFabric = MaskCharacterFabric()
    private var mPrepopulateCharacter: MutableList<MaskCharacter>? = null

    constructor(fmtString: String) : this() {
        formatString = fmtString
        mMask = buildMask(formatString)
    }

    fun size(): Int {
        return mMask.size
    }

    operator fun get(index: Int): MaskCharacter {
        return mMask[index]
    }


    fun isValidPrepopulateCharacter(ch: Char, at: Int): Boolean {
        return try {
            val character = mMask[at]
            character.isPrepopulate && character.isValidCharacter(ch)
        } catch (e: IndexOutOfBoundsException) {
            false
        }

    }

    fun isValidPrepopulateCharacter(ch: Char): Boolean {
        for (maskCharacter in mPrepopulateCharacter!!) {
            if (maskCharacter.isValidCharacter(ch)) {
                return true
            }
        }
        return false
    }


    private fun buildMask(fmtString: String): List<MaskCharacter> {
        val result = ArrayList<MaskCharacter>()
        mPrepopulateCharacter = ArrayList()
        for (ch in fmtString.toCharArray()) {
            val maskCharacter = mFabric.buildCharacter(ch)
            if (maskCharacter?.isPrepopulate == true) {
                mPrepopulateCharacter!!.add(maskCharacter)
            }
            result.add(maskCharacter!!)
        }
        return result
    }


    fun getFormattedString(value: String): IFormattedString {
        return FormattedString(this, value)
    }


}
